File: Practica 1/QuickSort/Python/QuickSort.py
from time import time
def partition(arr, low, high):
    i = (low-1)         
    pivot = arr[high]     
    for j in range(low, high):
        if arr[j] <= pivot:
            i = i+1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return (i+1)


def QuickSort(arr):
    def quickSort(arr, low, high):
        if len(arr) == 1:
            return arr
        if low < high:
            # at right place
            pi = partition(arr, low, high)
            quickSort(arr, low, pi-1)
            quickSort(arr, pi+1, high)
    quickSort(arr, 0, len(arr)-1)
    return arr

#Read data
numbers =[]
with open('D:/Cursos/Maestría CS/Algoritmos y Estructuras de Datos/Primera Práctica/QuickSort/Python/PuntosReal.txt','r') as file:
    for line in file:           
        for num in line.split():         
            numbers.append(int(num))

outputFile = open('D:/Cursos/Maestría CS/Algoritmos y Estructuras de Datos/Primera Práctica/QuickSort/Python/timesQuickSortPython.txt','w')

cases = 15
cont=0
iterations = numbers[cont]
cont=cont+1
for i in range (cases):    
    sizeArray=numbers[cont]
    cont=cont+1
    print(sizeArray, end=" ")
    outputFile.write(str(sizeArray)+" ")
    array = []
    for j in range (iterations):
        for k in range (sizeArray):
            array.append(numbers[cont])
            cont=cont+1            
        start = time()
        QuickSort(array)
        end = time()
        print(int((end-start)*1000000000), end=" ")
        outputFile.write(str(int((end-start)*1000000000))+" ")
    print("\n")
    outputFile.write("\n")

outputFile.close()

File: Practica 1/QuickSort/C++/QuickSort.cpp
#include <iostream>
#include <fstream>
#include <chrono>

int partition(int*, int , int );
void Quicksort(int* , int , int );
void print(int* ,int );
void generateTimes();

using namespace std;
int main() {
  generateTimes();
}

void Quicksort(int* arr, int low, int high) {
  if (low < high) {
    int pi = partition(arr, low, high);
    Quicksort(arr, low, pi - 1);
    Quicksort(arr, pi + 1, high);
  }
}

int partition(int* arr, int low, int high) {
  int pivot = arr[high];
  int i = (low - 1);  
  for (int j = low; j < high; j++) {
    if (arr[j] < pivot) {
      i++;
      int temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
    }
  }
  int temp = arr[i + 1];
  arr[i + 1] = arr[high];
  arr[high] = temp;
  return i + 1;
}

void generateTimes(){
  ifstream inputFile("PuntosReal.txt");
  ofstream outputFile("timesQuickSortC++.txt");
  int cases=15, iterations, sizeArray;
  inputFile>>iterations;  
  //Number o cases
  for(int i=0; i<cases;i++){
    inputFile>>sizeArray;
    outputFile<<sizeArray;
    int array[sizeArray];
    for(int j=0; j<iterations;j++){    
      for(int k=0;k<sizeArray;k++){
        inputFile>>array[k];
      }
      auto start = chrono::system_clock::now();
      Quicksort(array,0,sizeArray);
      auto end = chrono::system_clock::now();
		  chrono::duration<long,std::nano> duration = end-start;
      long time = duration.count();        
      outputFile<<" "<<time;
    }
    outputFile<<endl;    
  }
  inputFile.close();
  outputFile.close();  
}

void print(int* array,int size){
	for(int i=0;i<size;i++){
		cout<<array[i]<<" ";
	}
	cout<<endl;
}


